cbits/fpstring.c

```diff
@@ -44,7 +48,21 @@ void fps_intersperse(unsigned char *q,
                      unsigned char *p,
                      size_t n,
                      unsigned char c) {
-
+#if defined(__x86_64__)
+  {
+    const __m128i separator = _mm_set1_epi8(c);
+    const unsigned char *const p_begin = p;
+    const unsigned char *const p_end = p_begin + n - 9;
+    while (p < p_end) {
+      const __m128i eight_src_bytes = _mm_loadl_epi64((__m128i *)p);
```

Comment:
What happens when `p` is not 64-bit aligned?  Is there any performance degradation?
Are there OS platforms supported by GHC where `__x86_64__` is defined, but the intrinsics are not available?
That is I guess, has this code been tested on FreeBSD, MacOS and Windows, and perhaps even less common platforms like NetBSD, that are not officially supported by GHC, but do maintain ports...

Lastly, I am curious about what sort of applications might care about the performance of intersperse?  Is this a graphics thing? When would I want to bulk insert a fixed byte between every other byte in a large-enough buffer to want it done faster with intrinsics? (And yet be doing the task in Haskell...)

Reply:
> What happens when p is not 64-bit aligned? Is there any performance degradation?

My understanding is that there is no performance degradation on CPUs released in the latest ~10 years. And for earlier ones I suspect it's still worth it.

> Are there OS platforms supported by GHC where __x86_64__ is defined, but the intrinsics are not available? That is I guess, has this code been tested on FreeBSD, MacOS and Windows, and perhaps even less common platforms like NetBSD, that are not officially supported by GHC, but do maintain ports...

I *think* this code should work on all OSes, at least with gcc and clang. Does it ever happen that cbits of a Haskell package are compiled with MSVC? Intrinsics should still be the same but I'd need to verify that the header names are correct.

I can make a PR configuring GitHub Actions to build and test `bytestring` on Linux, Mac and Windows if you're OK with that. I'm not aware of any CI service that offers BSD machines.

> Lastly, I am curious about what sort of applications might care about the performance of intersperse? Is this a graphics thing? When would I want to bulk insert a fixed byte between every other byte in a large-enough buffer to want it done faster with intrinsics? (And yet be doing the task in Haskell...)

I just want the concept of using SIMD in Haskell to be more mainstream. The reason I started with `ByteString.intersperse` is that it's very similar to ascii->utf16 conversion that I accelerated in `text` here: https://github.com/haskell/text/pull/298. I'd like to continue with other functions like `ByteString.reverse`.

Reply:
> My understanding is that there is no performance degradation...

@ethercrow could you please add a benchmark for non-aligned bytestrings?

Reply:
@Bodigrim what's a good way to create one that's surely unaligned? Take a slice of global one with an odd offset like 7?

Reply:
Yeah, `Data.ByteString.drop 1` should be enough.